genie-web/src/main/java/com/netflix/genie/web/util/JobExecutionModeSelector.java

```diff
@@ -0,0 +1,233 @@
+
+    /**
+     * Property controlling default/fallback execution mode if none of the checks decides otherwise.
+     */
+    @VisibleForTesting
+    public static final String DEFAULT_EXECUTION_PROPERTY = PROPERTY_PREFIX + "default";
+
+    private final Environment environment;
+    private final Random random;
+    private final MeterRegistry meterRegistry;
+    private final List<Pair<String, CheckRequestFunction>> checks = ImmutableList.of(
+        // Global override that applies to all incoming jobs
+        new ImmutablePair<>("global-override", this::checkGlobalOverride),
+        // Job has HTTP header explicitly requesting a given execution mode
+        new ImmutablePair<>("header-override", this::checkForcingHeader),
+        // Job execution via agent for a fixed percentage of jobs, chosen randomly
+        new ImmutablePair<>("percentage", this::checkProbability)
+    );
+
+    /**
+     * Constructor.
+     *
+     * @param environment   the environment
+     * @param meterRegistry the meter registry
+     */
+    public JobExecutionModeSelector(
+        final Environment environment,
+        final MeterRegistry meterRegistry
+    ) {
+        this(new Random(), environment, meterRegistry);
+    }
+
+    /**
+     * Constructor with random generator used for tests.
+     *
+     * @param random        the random number generator
+     * @param environment   the environment
+     * @param meterRegistry the meter registry
+     */
+    @VisibleForTesting
+    JobExecutionModeSelector(
+        final Random random,
+        final Environment environment,
+        final MeterRegistry meterRegistry
+    ) {
+        this.environment = environment;
+        this.random = random;
+        this.meterRegistry = meterRegistry;
+    }
+
+
+    /**
+     * Inspect the incoming request and decide whether to execute using the new Agent-based execution mode.
+     *
+     * @param jobRequest         the job request
+     * @param httpServletRequest the raw http request
+     * @return true if this job should be executed using the agent
+     */
+    public boolean executeWithAgent(final JobRequest jobRequest, final HttpServletRequest httpServletRequest) {
+        // Perform checks in order
+        for (final Pair<String, CheckRequestFunction> nameFunctionCheckPair : this.checks) {
+            final String checkName = nameFunctionCheckPair.getKey();
+            final CheckRequestFunction checkFunction = nameFunctionCheckPair.getValue();
+            final Optional<Boolean> checkOptionalOutcome = checkFunction.check(jobRequest, httpServletRequest);
+
+            // Stop at the first check that returns a non-empty Optional
+            if (checkOptionalOutcome.isPresent()) {
+                final boolean executeWithAgent = checkOptionalOutcome.get();
+                this.publishMetric(executeWithAgent, checkName);
+                return executeWithAgent;
+            }
+        }
```

Comment:
Seems cleaner to do this in a `ifPresent` block than assigning it to a variable and then doing an `isPresent` check

Reply:
I cannot return in an `ifPresent` block (it returns from the lambda)